File: Dynamic Programing/Fibonacci/java/FibonacciInRecursionOptimized.java
import java.util.Scanner;

public class FibonacciInRecursionOptimized{
	public static int MAX_VALUE = 21;
	public static int f[] = new int[MAX_VALUE];
	public static int fib(int n){
		if(n == 0 || n == 1) return n;
		if(f[n] != -1) return f[n];
		
		f[n] = fib(n-1) + fib(n-2);
		return f[n];
	}
	public static void init(){
		for(int i=0; i<f.length; i++){
			f[i] = -1;
		}
	}
	public static void main(String[] args){
		Scanner input = new Scanner(System.in);
		System.out.println("Enter n'th number to find fibonacci number\nMax limit 20");
		int n;
		init();
		while(true){
			n = input.nextInt();
			if(n <=0 || n>=21) break;
			System.out.println(fib(n));
		}
	}
}

File: Dynamic Programing/Fibonacci/Python/fibonacci_in_recursion_optimized.py
memo = [-1]*20

def fib(n):
    if n == 0 or n == 1: return n

    if memo[n] != -1: return memo[n]

    memo[n] = fib(n-1) + fib(n-2)
    return memo[n]

while True:
    n = int(input())
    if n<= 0: break
    print(fib(n))

    
